### PR TITLE
`hvncat`: Added inbounds annotations that improve performance

### PR DESCRIPTION
**Pull request by BioTurboNick**
Breaking down #41143 into smaller pieces.

Moved `hvncat_fill!` method and added judicious inbounds/~inline~ annotations.

### CODE REVIEW

**Comment by simeonschaub** [comment]
Would it be possible to get some rough performance numbers for this?

**Comment by BioTurboNick** [comment]
Each exercises a different `hvncat_fill!` method
```
julia> const b = fill(1, 4, 4, 4);

julia> @btime [b b ;;; b b];
  1.080 μs (3 allocations: 2.34 KiB)    # inbounds and inlined
  1.350 μs (4 allocations: 2.39 KiB)    # inbounds and not inlined
  1.740 μs (4 allocations: 2.39 KiB)    # no inbounds and inlined
  1.720 μs (4 allocations: 2.39 KiB)    # no inbounds and not inlined

julia> @btime [1 2 3 4 5 6 7 8 9 ; 9 8 7 6 5 4 3 2 1 ;;; 5 5 5 5 5 5 5 5 5 ; 8 8 8 8 8 8 8 8 8]
  1.590 μs (6 allocations: 1.06 KiB)    # inbounds and inlined
  2.078 μs (7 allocations: 1.36 KiB)    # inbounds and not inlined
  1.580 μs (6 allocations: 1.06 KiB)    # no inbounds and inlined
  2.078 μs (7 allocations: 1.36 KiB)    # no inbounds and not inlined
```

File: base/abstractarray.jl [hunk 1]
```diff
@@ -2370,28 +2345,63 @@ function _typed_hvncat(::Type{T}, shape::Tuple{Vararg{Tuple, N}}, row_first::Boo
     return A
 end
 
-function hvncat_fill!(A::AbstractArray{T, N}, scratch1::Vector{Int}, scratch2::Vector{Int}, d1::Int, d2::Int, as::Tuple{Vararg}) where {T, N}
+@inline function hvncat_fill!(A::Array, row_first::Bool, xs::Tuple)
+    # putting these in separate functions leads to unnecessary allocations
+    lenxs = length(xs)
+    lena = length(A)
+    lenxs == lena || throw(ArgumentError("number of elements don't match specified shape"))
```
Comment (JeffBezanson) [inline base/abstractarray.jl:2352]:
```suggestion
    lenxs == lena || throw(ArgumentError("number of elements doesn't match specified shape"))
```

**Comment by JeffBezanson** [comment]
I don't think we should declare these `@inline`. That might make a microbenchmark of a single call faster, but it could potentially be very hard on the compiler to have to re-compile many copies of these; the functions are pretty large. Or if a large function has many uses of this, code size can become a problem, potentially even causing worse run-time performance.

**Comment by BioTurboNick** [comment]
> I don't think we should declare these `@inline`. That might make a microbenchmark of a single call faster, but it could potentially be very hard on the compiler to have to re-compile many copies of these; the functions are pretty large. Or if a large function has many uses of this, code size can become a problem, potentially even causing worse run-time performance.

Is there a way to see if that is an issue?

**Comment by JeffBezanson** [comment]
The burden of proof is on the other side; the correct default is not to declare it, and let the compiler decide.

**Comment by BioTurboNick** [comment]
Fair enough. Keep `@inbounds` for the one method that saw improvement? I suspect it might even just be the call to `hvncat_calcindex` that benefits.

**Comment by BioTurboNick** [comment]
Rebased. Still a performance improvement here with `@inbounds`:

```
julia> const b = fill(1, 4, 4, 4);

julia> e() = [b b ;;; b b]

julia> @btime e();
  1.300 μs (4 allocations: 2.33 KiB) # inbounds
  2.822 μs (4 allocations: 2.33 KiB) # no inbounds
```

@simeonschaub Could we please get this in?

**Comment by BioTurboNick** [comment]
@vtjnash @simeonschaub - Can this small PR please be merged for 1.9?

File: base/abstractarray.jl [hunk 2]
```diff
@@ -2583,28 +2583,35 @@ function _typed_hvncat_shape(::Type{T}, shape::NTuple{N, Tuple}, row_first, as::
...
+    d1 != d2 ||
+        throw(ArgumentError("d1 and d2 must be either 1 or 2, exclusive."))
     outdims = size(A)
     offsets = scratch1
     inneroffsets = scratch2
     for a ∈ as
         if isa(a, AbstractArray)
             for ai ∈ a
-                Ai = hvncat_calcindex(offsets, inneroffsets, outdims, N)
+                @inbounds Ai = hvncat_calcindex(offsets, inneroffsets, outdims, N)
```
Comment (dkarrasch) [inline base/abstractarray.jl:2600]: This line doesn't have indexing calls, so is `@inbounds` really effective here? Same below. Or do you wanna move it two lines up?

Comment (BioTurboNick) [reply base/abstractarray.jl:2600]: Yes, `hvncat_calcindex` is marked `@propagate_inbounds`

**Comment by BioTurboNick** [comment]
Bump!

**Comment by BioTurboNick** [comment]
bump

File: base/abstractarray.jl [hunk 3]
```diff
@@ -2609,28 +2609,35 @@ function _typed_hvncat_shape(::Type{T}, shape::NTuple{N, Tuple}, row_first, as::
     return A
 end
 
-function hvncat_fill!(A::AbstractArray{T, N}, scratch1::Vector{Int}, scratch2::Vector{Int}, d1::Int, d2::Int, as::Tuple{Vararg}) where {T, N}
+function hvncat_fill!(A::AbstractArray{T, N}, scratch1::Vector{Int}, scratch2::Vector{Int},
+                              d1::Int, d2::Int, as::Tuple) where {T, N}
+    length(scratch1) == length(scratch2) == N ||
+        throw(ArgumentError("scratch vectors must have as many elements as the destination array has dimensions"))
+    0 < d1 < 3 &&
```
Comment (Keno) [inline base/abstractarray.jl:2616]:
```suggestion
    0 < d1 <= N &&
```
?
Otherwise the inbounds below is out of bounds

**Review by Keno — APPROVED** [review]
LGTM with comment addressed

File: base/abstractarray.jl (hunk 3 above)
Comment (BioTurboNick) [reply base/abstractarray.jl:2616]: d1 has to be 1 or 2, and this function is only called for N > 2. Dispatch takes care of that, though I can add a check to ensure `N > 2` to be extra safe.

File: base/abstractarray.jl [hunk 4]
```diff
@@ -2422,7 +2422,7 @@ function hvncat_fill!(A::Array, row_first::Bool, xs::Tuple)
             end
         end
     else
-        for k ∈ eachindex(xs)
+        @inbounds for k ∈ eachindex(xs)
             A[k] = xs[k]
```
Comment (giordano) [inline base/abstractarray.jl:2426]: Would
```suggestion
        for k ∈ eachindex(A, xs)
            A[k] = xs[k]
```
let us avoid `@inbounds` and retain same performance?

**Review by vtjnash — COMMENTED** [review]
I am a little worried about marking all of those `iterate` calls being `@inbounds`. This seems like rather a large amount of code to mark `@inbounds` on each block, and `hvncat_fill!` appears to fail to properly boundscheck `xs[k]`, for example.

File: base/abstractarray.jl (hunk 4 above)
Comment (BioTurboNick) [reply base/abstractarray.jl:2426]: Wasn't aware of the multi-arg `eachindex`, but looks like no:
```
julia> eachindex([1, 2, 3], (2, 3, 4))
ERROR: MethodError: no method matching keys(::Vector{Int64}, ::Tuple{Int64, Int64, Int64})
```

**Comment by BioTurboNick** [comment]
> `hvncat_fill!` appears to fail to properly boundscheck `xs[k]`, for example.

Fair, easy enough to address.

**Comment by BioTurboNick** [comment]
> I am a little worried about marking all of those `iterate` calls being `@inbounds`. This seems like rather a large amount of code to mark `@inbounds` on each block

I'll check and narrow it if it still works - was defaulting to how `hvcat_fill!` does it.

**Comment by BioTurboNick** [comment]
All green

**Pull request merged**